src/meltano/core/meltano_file.py

```diff
@@ -47,6 +50,69 @@ def __init__(
             jobs=self.load_job_tasks(jobs or []),
         )
 
+        self.expansion_env = dict(os.environ)
+        self.base_plugin_environments = defaultdict(dict)
+
+        # Expand env vars in top level based on terminal context
+        if "env" in self.extras:
+            self.extras.update(
+                {
+                    "env": expand_env_vars(
+                        self.extras["env"], self.expansion_env, recurse=True
+                    )
+                }
+            )
+            self.expansion_env.update(self.extras["env"])
+
+        # Expand env vars in environments based on terminal context + top level
+        for environment in self.environments:
+            environment.env.update(
+                expand_env_vars(environment.env, self.expansion_env, recurse=True)
+            )
+            if environment.name == os.environ.get(
+                "MELTANO_ENVIRONMENT", self.default_environment
+            ):
+                self.expansion_env.update(environment.env)
+
+        # Expand env vars in plugins based on terminal context + top level
+        # + active environment
+        for plugin_type, plugins_of_type in self.plugins:
+            for plugin in plugins_of_type:
+                if "env" in plugin.extras:
+                    plugin.extras["env"].update(
+                        expand_env_vars(
```

Comment:
Most of what we're doing here seems like it would be more appropriately done in `(Project/Plugin)SettingsService` when we're putting together the env to use to read setting values and handle expansion, by having a clear list of layers of envs, instead of updating `env` properties in place.